server/channels/app/notification.go

```diff
@@ -1214,68 +1215,108 @@ func (a *App) FilterUsersByVisible(c request.CTX, viewer *model.User, otherUsers
 	}
 
 	mentionedUsersInTheTeam, err := a.Srv().Store().User().GetProfilesByUsernames(potentialMentions, &model.ViewUsersRestrictions{Teams: []string{channel.TeamId}})
 	if err != nil {
-		return nil, nil, nil, err
+		return nil, nil, nil, nil, err
 	}
 
 	// Filter out inactive users and bots
 	teamUsers := model.UserSlice(mentionedUsersInTheTeam).FilterByActive(true)
 	teamUsers = teamUsers.FilterWithoutBots()
 	teamUsers, appErr := a.FilterUsersByVisible(c, sender, teamUsers)
 	if appErr != nil {
-		return nil, nil, nil, appErr
+		return nil, nil, nil, nil, appErr
 	}
 
 	allMentionedUsers, err := a.Srv().Store().User().GetProfilesByUsernames(potentialMentions, nil)
 	if err != nil {
-		return nil, nil, nil, err
+		return nil, nil, nil, nil, err
 	}
 
 	outOfTeamUsers := model.UserSlice(allMentionedUsers).FilterWithoutID(teamUsers.IDs())
 	outOfTeamUsers = outOfTeamUsers.FilterByActive(true)
 	outOfTeamUsers = outOfTeamUsers.FilterWithoutBots()
 	outOfTeamUsers, appErr = a.FilterUsersByVisible(c, sender, outOfTeamUsers)
 	if appErr != nil {
-		return nil, nil, nil, appErr
+		return nil, nil, nil, nil, appErr
 	}
 
 	if len(teamUsers) == 0 {
-		return outOfTeamUsers, nil, nil, nil
+		return outOfTeamUsers, nil, nil, nil, nil
 	}
 
 	// Differentiate between mentionedUsersInTheTeam who can and can't be added to the channel
 	var outOfChannelUsers model.UserSlice
 	var outOfGroupsUsers model.UserSlice
-	if channel.IsGroupConstrained() {
-		nonMemberIDs, err := a.FilterNonGroupChannelMembers(teamUsers.IDs(), channel)
-		if err != nil {
-			return nil, nil, nil, err
-		}
+	var nonInvitableUsers model.UserSlice
+
+	// Check if channel is ABAC controlled and filter users accordingly
+	if ok, err := a.ChannelAccessControlled(c, channel.Id); err != nil {
+		return nil, nil, nil, nil, err
+	} else if ok {
+		// Channel is ABAC controlled - filter users based on attributes
+		acs := a.Srv().ch.AccessControl
+		if acs != nil {
+			// Get all users that can potentially access this channel
+			filteredUsers, _, appErr := acs.QueryUsersForResource(c, channel.Id, "*", model.SubjectSearchOptions{
+				Term:   "",
+				TeamID: channel.TeamId,
+				Limit:  1000, // Set a reasonable limit
+			})
+			if appErr != nil {
+				return nil, nil, nil, nil, appErr
+			}
+
+			// Create a map for quick lookup of users who match ABAC attributes
+			allowedUserIDs := make(map[string]bool)
+			for _, user := range filteredUsers {
+				allowedUserIDs[user.Id] = true
+			}
 
-		outOfChannelUsers = teamUsers.FilterWithoutID(nonMemberIDs)
-		outOfGroupsUsers = teamUsers.FilterByID(nonMemberIDs)
+			// Separate users into those who can and cannot be invited
+			for _, user := range teamUsers {
+				if allowedUserIDs[user.Id] {
+					outOfChannelUsers = append(outOfChannelUsers, user)
+				} else {
+					nonInvitableUsers = append(nonInvitableUsers, user)
+				}
+			}
+		} else {
+			// ABAC service not available, treat all as invitable
+			outOfChannelUsers = teamUsers
+		}
```

Comment:
Can you make this a separate method that returns `nonInvitableUsers` and an error? It can be easier to read and understand the code.

We could retain the existing flow but just add the new value accordingly. The new function will just return `nil`, `nil` if the channel is not abac enforced or the license doesn't permit etc.